### PR TITLE
Added devcontainer

### PR DESCRIPTION
**Pull request by WorksButNotTested**

### CODE REVIEW

**Comment by wsxiaoys** [comment]
Please rebase against current main

File: Dockerfile [hunk 1]
```diff
@@ -7,6 +7,9 @@ ARG BASE_CUDA_DEV_CONTAINER=nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_VER
 ARG BASE_CUDA_RUN_CONTAINER=nvidia/cuda:${CUDA_VERSION}-runtime-ubuntu${UBUNTU_VERSION}
 
 FROM ${BASE_CUDA_DEV_CONTAINER} as build
+ENV CTRANSLATE2_ROOT=/opt/ctranslate2
+COPY --from=source $CTRANSLATE2_ROOT $CTRANSLATE2_ROOT
+ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CTRANSLATE2_ROOT/lib
```
Comment (wsxiaoys) [inline Dockerfile:12]:
```suggestion
ENV CTRANSLATE2_ROOT=/opt/ctranslate2
COPY --from=source $CTRANSLATE2_ROOT $CTRANSLATE2_ROOT
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CTRANSLATE2_ROOT/lib
```

These are no longer needed as tabby switched to llama.cpp for cuda

File: Dockerfile [hunk 2]
```diff
@@ -25,6 +28,9 @@ RUN apt-get update && \
 RUN curl https://sh.rustup.rs -sSf | bash -s -- -y
 ENV PATH="/root/.cargo/bin:${PATH}"
 
+RUN git config --system --add safe.directory "*"
+
+FROM builder as built
```
Comment (wsxiaoys) [inline Dockerfile:33]:
```suggestion
FROM builder as rust_build
```

File: Dockerfile [hunk 3]
```diff
@@ -55,7 +61,7 @@ RUN git config --system --add safe.directory "*"
 RUN ln -s /usr/lib/x86_64-linux-gnu/libnvidia-ml.so.1 \
         /usr/lib/x86_64-linux-gnu/libnvidia-ml.so
 
-COPY --from=build /opt/tabby /opt/tabby
+COPY --from=built /opt/tabby /opt/tabby
```
Comment (wsxiaoys) [inline Dockerfile:64]:
```suggestion
COPY --from=rust_build /opt/tabby /opt/tabby
```

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{
+	"name": "TabbyML",
+	"build": {
+		"context": "..",
+		"dockerfile": "../Dockerfile",
+		"target": "builder"
```
Comment (wsxiaoys) [inline .devcontainer/devcontainer.json:6]: should be `build` here?

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (WorksButNotTested) [reply Dockerfile]: Done

File: Dockerfile (hunk 3 above)
Comment (WorksButNotTested) [reply Dockerfile]: Done

File: .devcontainer/devcontainer.json (hunk 4 above)
Comment (WorksButNotTested) [reply .devcontainer/devcontainer.json]: Done

File: Dockerfile (hunk 2 above)
Comment (WorksButNotTested) [reply Dockerfile]: Done

**Comment by WorksButNotTested** [comment]
Should be addressed now.

**Pull request closed without merging**